python/lsst/rubintv/production/starTracker.py

```diff
@@ -642,3 +676,165 @@ def callback(self, filename, doCheckDay=True):
+        self.log.info(f'of which {len(filenames)} had corresponding files')
+
+        # send a heartbeat saying we'll be back in a while
+        maxProcessingTimePerFile = 30  # a deliberate massive over-estimate, in seconds
+        forecast = maxProcessingTimePerFile * len(filenames)
+        self.heartbeater.beat(customFlatlinePeriod=forecast)
+
+        starTrackerChannel = StarTrackerChannel(locationConfig=self.locationConfig,
+                                                cameraType=camera.cameraType)
+        for file in filenames:
+            self.log.info(f'Catching up {file}')
+            try:
+                starTrackerChannel.callback(file)
+            except Exception as e:
+                raiseIf(self.doRaise, e, self.log)
+
+    def runEndOfDayManual(self, dayObs):
+        """Manually run the end of day routine for a specific dayObs by hand.
+
+        Useful for if the final catchup and end of day animation/clearup have
+        failed to run and this needs to be redone by manually.
+
+        Parameters
+        ----------
+        dayObs : `int`
+            The dayObs to rerun the end of day routine for.
+        """
+        self.dayObs = dayObs
+        self.runCatchup()
+        return
+
+    def runCatchup(self):
+        """Run the catchup for all cameras.
+        """
+        for camera in self.cameras:
+            self.log.info(f'Starting catchup for the {camera.cameraType} camera')
+            self.catchupCamera(camera)
+
+    def runEndOfDay(self):
+        """Routine to run when the summit dayObs rolls over.
+
+        Sets the new dayObs.
+        """
+        try:
+            self.runCatchup()
+        except Exception as e:
+            raiseIf(self.doRaise, e, self.log)
+        finally:
+            self.dayObs = getCurrentDayObs_int()
+
+    def run(self):
+        """Runs forever, running the catchup services during the night, and
+        rolls the dayObs over at the end of the night.
+
+        Raises
+        ------
+        RuntimeError:
+            Raised from the root error on any error if ``self.doRaise`` is
+            True.
+        """
+        lastRun = time.time()
+        self.dayObs = getCurrentDayObs_int()
+
+        while True:
+            try:
+                timeSince = time.time() - lastRun
+                if timeSince >= self.catchupPeriod:
+                    self.runCatchup()
+                    self.heartbeater.beat()
+                    lastRun = time.time()
+                    if hasDayRolledOver(self.dayObs):
+                        sleep(self.endOfDayDelay)  # give time for anything running elsewhere to finish
+                        self.runEndOfDay()  # sets new dayObs in a finally block
+                else:
+                    remaining = self.catchupPeriod - timeSince
+                    self.log.info(f'Waiting for catchup period to elapse, {remaining:.2f}s to go...')
+                    sleep(self.loopSleep)
```

Comment:
Should this sleep for the larger of `remaining` and `loopSleep`?  Or is this potentially shorter to ensure that the heartbeat is sent on line 837?

Reply:
Yeah, to be honest all that sleeping stuff is pretty rough, and just makes sure we don't unnecessarily hammer things. The point isn't to count all the way down in one go though, for the exact reason you said.